pkg/workflow/controllers/serial_node_reconciler.go

```diff
@@ -0,0 +1,144 @@
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// only resolve serial nodes
+	if node.Spec.Type != v1alpha1.TypeSerial {
+		return reconcile.Result{}, nil
+	}
+
+	// empty serial node
+	if len(node.Spec.Tasks) == 0 {
+		it.logger.V(4).Info("empty serial node, NOOP", "key", request.NamespacedName)
+		return reconcile.Result{}, nil
+	}
+
+	// un-synced expected children
+	if node.Status.ExpectedChildren == nil {
+		expected := len(node.Spec.Tasks)
+		node.Status.ExpectedChildren = &expected
+	}
+
+	// this node should finished
+	if len(node.Status.FinishedChildren) == *node.Status.ExpectedChildren {
+		if !ConditionEqualsTo(node.Status, v1alpha1.ConditionAccomplished, corev1.ConditionTrue) {
+			updateError := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				node := v1alpha1.WorkflowNode{}
+				err := it.kubeClient.Get(ctx, request.NamespacedName, &node)
+				if err != nil {
+					return client.IgnoreNotFound(err)
+				}
+				SetCondition(&node.Status, v1alpha1.WorkflowNodeCondition{
+					Type:   v1alpha1.ConditionAccomplished,
+					Status: corev1.ConditionTrue,
+					Reason: v1alpha1.NodeAccomplished,
+				})
+				return it.kubeClient.Update(ctx, &node)
+			})
+
+			if updateError != nil {
+				return reconcile.Result{}, updateError
+			}
+
+			it.eventRecorder.Event(&node, corev1.EventTypeNormal, v1alpha1.NodeAccomplished, "Serial node accomplished")
+		}
+		return reconcile.Result{}, nil
+	}
+
+	if len(node.Status.ActiveChildren) == 0 {
+		it.logger.Info("schedule next", "key", request.NamespacedName, "status", node.Status)
+		taskToStartup := node.Spec.Tasks[len(node.Status.FinishedChildren)]
+		parentWorkflow := v1alpha1.Workflow{}
+
+		err := it.kubeClient.Get(ctx, types.NamespacedName{
+			Namespace: node.Namespace,
+			Name:      node.Spec.WorkflowName,
+		}, &parentWorkflow)
+		if err != nil {
+			it.logger.Error(err, "failed to fetch parent workflow", "key", request.NamespacedName, "workflow name", node.Spec.WorkflowName)
+			return reconcile.Result{}, err
+		}
+		childrenNodes, err := renderNodesByTemplates(&parentWorkflow, &node, taskToStartup)
+		if err != nil {
+			it.logger.Error(err, "failed to render children childrenNodes", "node", request.NamespacedName)
+			return reconcile.Result{}, err
+		}
+
+		for _, childNode := range childrenNodes {
+			err := it.kubeClient.Create(ctx, childNode)
+			if err != nil {
+				it.logger.Error(err, "failed to create child node", "node", request.NamespacedName, "child node", childNode)
+				return reconcile.Result{}, err
+			}
+		}
+
+		updateError := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			node := v1alpha1.WorkflowNode{}
+			err := it.kubeClient.Get(ctx, request.NamespacedName, &node)
+			if err != nil {
+				return err
+			}
+			for _, item := range childrenNodes {
+				node.Status.ActiveChildren = append(node.Status.ActiveChildren, corev1.LocalObjectReference{Name: item.Name})
+			}
+			return it.kubeClient.Update(ctx, &node)
```

Comment:
Yes, it might make a mess with the high-presssure situation. 

The reason is now I do not make "create node" and "update state" as an atomic operation. I think I could fix that by "labeling when creating" and use "list with label selector" to resolve that.

I think I should resolve it in another PR.🤔 Currently, it could work for most situations.